src/main/java/com/cburch/logisim/proj/Project.java

```diff
@@ -37,9 +37,60 @@
 import java.util.HashMap;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.ArrayList;
+import java.util.Iterator;
 import javax.swing.JFileChooser;
 
 public class Project {
+  public void discardAllEdits() {
+    undoLog.clear();
+    redoLog.clear();
+    undoMods = 0;
+    fireEvent(new ProjectEvent(ProjectEvent.ACTION_COMPLETE, this, null));
+  }
+
+  public List<Action> getRedoActions() {
+    List<Action> actions = new ArrayList<>();
+    Iterator<ActionData> iter = redoLog.descendingIterator();
+    while (iter.hasNext()) {
+      actions.add(iter.next().action);
+    }
+    return actions;
+  }
+
+  public void redoUpTo(Action targetAction) {
+    if (redoLog.isEmpty()) return;
+    Action lastRedoneAction = null;
+    do {
+      redoAction();
+      if (!undoLog.isEmpty()) {
+        lastRedoneAction = undoLog.getLast().action;
+      }
+    } while (lastRedoneAction != targetAction && !redoLog.isEmpty());
+  }
+
+  public List<Action> getUndoActions() {
+    List<Action> actions = new ArrayList<>();
+    Iterator<ActionData> iter = undoLog.descendingIterator();
+    while (iter.hasNext()) {
+      actions.add(iter.next().action);
+    }
+    return actions;
+  }
+
+  public void undoUpTo(Action targetAction) {
+    if (undoLog.isEmpty()) return;
+    Action lastUndoneAction = null;
+    do {
+      if (undoLog.isEmpty()) break;
+      undoAction();
+      if (!redoLog.isEmpty()) {
+        lastUndoneAction = redoLog.getLast().action;
+      }
+    } while (lastUndoneAction != targetAction && !undoLog.isEmpty());
```

Comment:
This loop might be better written as a regular while loop rather than a do-while loop. If you use the same while condition, you will not need the `if (undoLog.isEmpty()) break;` in the loop. You also don't need the first if in the method.

Note that the same might also be applied to the redoUpTo method.